blenderproc/python/types/URDFUtility.py

```diff
@@ -0,0 +1,898 @@
+        object.__setattr__(self, "ik_bone_constraint", bone)
+
+    def _set_fk_ik_mode(self, mode: str = "fk"):
+        object.__setattr__(self, "fk_ik_mode", mode)
+
+    def _set_ik_link(self, ik_link: Union[Link, None]):
+        object.__setattr__(self, "ik_link", ik_link)
+
+    def create_ik_bone_controller(self, link: Union[Link, None] = None,
+                                  relative_location: Union[List[float], Vector] = [0., 0., 0.], use_rotation: bool = True,
+                                  chain_length: int = 0):
+        """ Creates an ik bone controller and a corresponding constraint bone for the respective link.
+
+        :param link: The link to create an ik bone for. If None, will use the last link.
+        :param relative_location: Relative location of the ik bone controller w.r.t. the bone's location. This can be
+                                  used to shift the point of control further away from the end effector.
+        :param use_rotation: Whether to rotate the child links as well. Defaults to True.
+        :param chain_length: The number of parent links which are influenced by this ik bone. Defaults to 0 for all
+                             parents.
+        """
+        if self.ik_bone_controller is not None:
+            raise NotImplementedError("URDFObject already has an ik bone controller. More than one ik controllers are "
+                                      "currently not supported!")
+        if link is None:
+            link = self.links[-1]
+        ik_bone_controller, ik_bone_constraint, offset = link._create_ik_bone_controller(
+            relative_location=relative_location, use_rotation=use_rotation, chain_length=chain_length)
+        self._set_ik_bone_controller(ik_bone_controller)
+        self._set_ik_bone_constraint(ik_bone_constraint)
+        self._set_ik_bone_offset(offset=offset)
+        self._set_ik_link(link)
+        self._switch_fk_ik_mode(mode="ik")
+
+    def _switch_fk_ik_mode(self, mode: str = "fk", keep_pose: bool = True):
+        """ Switches between forward and inverse kinematics mode. Will do this automatically when switching between e.g.
+            `set_rotation_euler_fk()` and `set_rotation_euler_ik()`.
+
+        :param mode: One of "fk", "ik".
+        :param keep_pose: If specified, will keep the pose when switching modes. Otherwise will return to the old pose
+                          of the previously selected mode.
+        """
+        if mode == "ik" and self.ik_bone_controller is None:
+            raise NotImplementedError("URDFObject doesn't have an ik bone controller. Please set up an ik bone first "
+                                      "with 'urdf_object.create_ik_bone_controller()'")
+        if self.fk_ik_mode != mode:
+            for link in self.links:
+                link.switch_fk_ik_mode(mode=mode, keep_pose=keep_pose)
+            self._set_fk_ik_mode(mode=mode)
+
+    def get_revolute_joints(self) -> List[Link]:
```

Comment:
maybe `get_links_with_revolute_joints()` ?

Reply:
docu missing as well

Reply:
It would be also nice to have a function to get all (revolute) joint angles at a given keyframe. That way you could easily log forward kinematics and reproduce trajectories. Something along:

```python
    def get_joint_rotation(self, frame: int = None):
        """ Get current joint rotation based on euler angles.
        :return: Current joint rotation in radians
        """
        bone = self.bone
        bpy.ops.object.select_all(action='DESELECT')
        bone.bone.select = True
        bone.rotation_mode = 'XYZ'
        with KeyFrame(frame):
            current_rotation_euler = bone.rotation_euler
        
        axis = self._determine_rotation_axis(bone=bone, allow_fixed=True)
        if axis is None:
            return 0.0
        return current_rotation_euler[["X", "Y", "Z"].index(axis)]
```

Reply:
agree